### PR TITLE
RPC-114-burn-1155

### PR DESCRIPTION
**Pull request by SKonstantinS**
For tokens 1155 fixes:
1.burn before mint
2.case: mint = 5, burn =7

### CODE REVIEW

File: tokens/test/erc-1155/ERC1155Rarible.test.js [hunk 1]
```diff
@@ -30,6 +30,485 @@ contract("ERC1155Rarible", accounts => {
...
+      const tokenId4 = minter + "b00000000000000000000004"; //save token creator
+      const tokenURI = "/uri";
+      let supply = 10;
+      let burn = 7;
+      let mintValue = 5;
+      //mint, after do burnBatch
+      await token.mintAndTransfer([tokenId4, tokenURI, supply, creators([minter]), [], [zeroWord]], transferTo, mintValue, {from: minter})
+      assert.equal(await token.balanceOf(transferTo, tokenId4), mintValue);
+      await expectThrow(  //tokenId4 send to transferTo, try to burn from anotherUser, throw
+        token.burnBatch(anotherUser, [tokenId4], [mintValue], {from: anotherUser})
```
Comment (disenotov) [inline tokens/test/erc-1155/ERC1155Rarible.test.js:121]: we should probably try to send this transaction from transferTo to check the specific require

File: tokens/test/erc-1155/ERC1155Rarible.test.js [hunk 2]
```diff
@@ -30,6 +30,485 @@ contract("ERC1155Rarible", accounts => {
...
+      const tokenId1 = minter + "b00000000000000000000001"; //save token creator
+      const tokenId2 = minter + "b00000000000000000000002"; //save token creator
+      const tokenId3 = minter + "b00000000000000000000003"; //save token creator
+      const tokenURI = "/uri";
+      let supply = 5;
+      let burn = 2;
+      let mintValue = 5;
+      //BurnBatch and after do mint
+      await expectThrow(
+        token.burnBatch(anotherUser, [tokenId1, tokenId2, tokenId3], [burn, burn, burn], {from: anotherUser})  //token has another creator
```
Comment (disenotov) [inline tokens/test/erc-1155/ERC1155Rarible.test.js:91]: we should probably try to send this transaction from minter to check the specific require

**Review by disenotov — CHANGES_REQUESTED** [review]
some minor changes needed, the only real change is about setDefaultApproval()

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 3]
```diff
@@ -14,6 +14,10 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
     string public name;
     string public symbol;
 
+    function setDefaultApproval(address operator, bool hasApproval) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol:17]: done

File: tokens/contracts/erc-1155/ERC1155Lazy.sol [hunk 4]
```diff
@@ -133,5 +133,25 @@ abstract contract ERC1155Lazy is IERC1155LazyMint, ERC1155BaseURI, Mint1155Valid
...
+        } else {
+            return true;
+        }
+    }
+
+    function _setBurned(uint256 tokenId, uint amount) internal {
+        minted[tokenId] += amount;
+    }
+
+    function _getMinted(uint256 tokenId) internal returns (uint) {
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Lazy.sol:148]: done

File: tokens/contracts/erc-1155/ERC1155Lazy.sol [hunk 5]
```diff
@@ -133,5 +133,25 @@ abstract contract ERC1155Lazy is IERC1155LazyMint, ERC1155BaseURI, Mint1155Valid
...
+        return minted[tokenId];
+    }
+
+    function _getSupply(uint256 tokenId) internal returns (uint) {
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Lazy.sol:152]: done

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 6]
```diff
@@ -35,5 +39,60 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
...
+        require(
+            account == _msgSender() || isApprovedForAll(account, _msgSender()),
+            "ERC1155: caller is not owner nor approved"
+        );
+        address operator = _msgSender();
+
+        _beforeTokenTransfer(operator, account, address(0), ids, amounts, "");
+
+        for (uint i = 0; i < ids.length; i++) {
+            if (_isExist(ids[i])) {
```
[the remark this answers is not in the document]
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol:88]: good Idea!!! done

File: tokens/test/erc-1155/ERC1155Rarible.test.js (hunk 1 above)
Comment (SKonstantinS) [reply tokens/test/erc-1155/ERC1155Rarible.test.js:121]: This transaction run later on Line 140 and if burnAmount <= lazyAmount, result ok (minter can delete only lazy)/
Add check burnAmount > lazyAmount, throw Line 136

File: tokens/test/erc-1155/ERC1155Rarible.test.js (hunk 2 above)
Comment (SKonstantinS) [reply tokens/test/erc-1155/ERC1155Rarible.test.js:91]: ok, add this check

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 7]
```diff
@@ -35,5 +35,56 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
...
+        require(account == _msgSender(), "ERC1155: caller is not burner");
+        address minter = address(id >> 96);
+        require(minter == _msgSender(), "ERC1155: caller is not token owner");
+        ERC1155Lazy._setBurned(id, value);
+    }
+
+    function _burnExisted(address account, uint256 id, uint256 value) internal {
+        address owner = address(id >> 96);
+        if (owner == _msgSender()) {
+            uint256 balance = ERC1155Lazy._getSupply(id) - ERC1155Lazy._getMinted(id);
```
Comment (SherbakovaMA) [inline tokens/contracts/erc-1155/ERC1155Base.sol:56]: it's balance of lazy. "balanceLazy" is better for variable name

Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol:56]: agree, do it

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 8]
```diff
@@ -35,5 +35,56 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
...
+
+    function _burnLazy(address account, uint256 id, uint256 value) internal {
+        require(account == _msgSender(), "ERC1155: caller is not burner");
+        address minter = address(id >> 96);
+        require(minter == _msgSender(), "ERC1155: caller is not token owner");
+        ERC1155Lazy._setBurned(id, value);
+    }
+
+    function _burnExisted(address account, uint256 id, uint256 value) internal {
+        address owner = address(id >> 96);
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:54]: creator?

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 9]
```diff
@@ -35,5 +35,56 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
         return super.supportsInterface(interfaceId);
     }
 
+    function burn(address account, uint256 id, uint256 value) public virtual override {
+        if (_isExist(id)) {
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:39]: do we really need this if?
it seems logic here can be simplified very much

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 10]
```diff
@@ -35,5 +35,56 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
...
+            _burnExisted(account, id, value);
+        } else {
+            _burnLazy(account, id, value);
+        }
+    }
+
+    function _burnLazy(address account, uint256 id, uint256 value) internal {
+        require(account == _msgSender(), "ERC1155: caller is not burner");
+        address minter = address(id >> 96);
+        require(minter == _msgSender(), "ERC1155: caller is not token owner");
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:49]: do we need this condition? it's checked in _burnExisted

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 11]
```diff
@@ -35,5 +35,56 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
...
 
+    function burn(address account, uint256 id, uint256 value) public virtual override {
+        if (_isExist(id)) {
+            _burnExisted(account, id, value);
+        } else {
+            _burnLazy(account, id, value);
+        }
+    }
+
+    function _burnLazy(address account, uint256 id, uint256 value) internal {
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:46]: It's better to move auxiliary function to the bottom. more important high-level functions go to the top

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: tokens/contracts/erc-1155/ERC1155Base.sol (hunk 8 above)
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol]: agree, creator

File: tokens/contracts/erc-1155/ERC1155Base.sol (hunk 10 above)
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol]: agree simplify logic

File: tokens/contracts/erc-1155/ERC1155Base.sol (hunk 11 above)
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol]: agree, do it

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 12]
```diff
@@ -31,8 +60,27 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
...
+        uint supply = ERC1155Lazy._getSupply(id);
+        if (creator == _msgSender() && supply == 0) {
+            //token not exists, burn Lazy by creator
+            ERC1155Lazy._setBurned(id, value);
+            return 0;
+        }
+        burnResidue = value;
+        if (creator == _msgSender()) {
+            //calculate Lazy value available for burn
+            uint256 balanceLazy = supply - ERC1155Lazy._getMinted(id);
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:74]: Safe Math?
Also, supply can be less than minted?
I think yes

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 13]
```diff
@@ -31,8 +60,27 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
         return _tokenURI(id);
     }
 
-    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC1155Lazy, ERC165Upgradeable) returns (bool) {
-        return super.supportsInterface(interfaceId);
+    function _burnLazy(address account, uint256 id, uint256 value) internal returns (uint256 burnResidue) {
+        address creator = address(id >> 96);
+        uint supply = ERC1155Lazy._getSupply(id);
+        if (creator == _msgSender() && supply == 0) {
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:66]: Do we really need this if?

File: tokens/contracts/erc-1155/ERC1155Base.sol [hunk 14]
```diff
@@ -18,6 +18,35 @@ abstract contract ERC1155Base is OwnableUpgradeable, ERC1155DefaultApproval, ERC
         return ERC1155DefaultApproval.isApprovedForAll(_owner, _operator);
     }
 
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC1155Lazy, ERC165Upgradeable) returns (bool) {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function burn(address account, uint256 id, uint256 value) public virtual override {
+        uint256 burnMinted = _burnLazy(account, id, value);
```
Comment (evgenynacu) [inline tokens/contracts/erc-1155/ERC1155Base.sol:26]: Is it good name for variable?
What does it mean?
Is it "left to burn"?

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: tokens/contracts/erc-1155/ERC1155Base.sol (hunk 14 above)
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol]: change to leftToBurn

File: tokens/contracts/erc-1155/ERC1155Base.sol (hunk 13 above)
Comment (SKonstantinS) [reply tokens/contracts/erc-1155/ERC1155Base.sol]: delete this if, simplify code

**Pull request merged**